### ISSUE TITLE
How to use more than 40 devices! I have 100 sets of equipment! Unable to start

### ISSUE THREAD

**Issue by q6654282** [Genymobile/gnirehtet#464]
How to use more than 40 devices! I have 100 sets of equipment! Unable to start

**Comment by rom1v** [comment]
> Unable to start

What do you execute? What is the error message in the console?

**Comment by q6654282** [comment]
Memory allocation exceeded

**Comment by q6654282** [comment]
2022-10-29 13:14:23.902 ERROR AdbMonitor: Failed to monitor adb devices: Packet size should not be that big: 2222
2022-10-29 13:14:23.902 INFO AdbMonitor: Restarting adb daemon

**Comment by q6654282** [comment]
Problems! As above

**Comment by rom1v** [comment]
> How to use more than 40 devices! I have 100 sets of equipment! Unable to start
>
> Memory allocation exceeded

Yes, this is a known design issue (I can't find another related bug but I remember it exists): a buffer is allocated upfront for each connection (so there is no memory allocation during forwarding).

It should be rewritten to dynamically allocate/reuse packets.


> 2022-10-29 13:14:23.902 ERROR AdbMonitor: Failed to monitor adb devices: Packet size should not be that big: 2222

Indeed, the adb monitor does not expect more than 1024 bytes per read. To be fixed also in [`autoadb`](https://github.com/rom1v/autoadb).

Also refs #396.

**Comment by q6654282** [comment]
> > 如何使用超过 40 台设备！我有100套设备！无法启动
> > 超出内存分配
> 
> 是的，这是一个已知的设计问题（我找不到另一个相关的错误，但我记得它存在）：为每个连接预先分配了一个缓冲区（因此在转发期间没有内存分配）。
> 
> 它应该被重写以动态分配/重用数据包。
> 
> > 2022-10-29 13:14:23.902 错误 AdbMonitor：无法监控 adb 设备：数据包大小不应该那么大：2222
> 
> 实际上，adb 监视器不希望每次读取超过 1024 字节。也被固定在[`autoadb`](https://github.com/rom1v/autoadb).
> 
> 也参考#396。

yes! This is the problem of automatic cache allocation! Is it possible to upgrade the version to solve this problem! sure

**Comment by q6654282** [comment]
> > How to use more than 40 devices! I have 100 sets of equipment! Unable to start
> > Memory allocation exceeded
> 
> Yes, this is a known design issue (I can't find another related bug but I remember it exists): a buffer is allocated upfront for each connection (so there is no memory allocation during forwarding).
> 
> It should be rewritten to dynamically allocate/reuse packets.
> 
> > 2022-10-29 13:14:23.902 ERROR AdbMonitor: Failed to monitor adb devices: Packet size should not be that big: 2222
> 
> Indeed, the adb monitor does not expect more than 1024 bytes per read. To be fixed also in [`autoadb`](https://github.com/rom1v/autoadb).
> 
> Also refs #396.

yes! This is the problem of automatic cache allocation! Is it possible to upgrade the version to solve this problem!

**Comment by q6654282** [comment]
Is there a solution to this problem

**Comment by rom1v** [comment]
Not really without major design changes, that I don't plan to make because I'm not really working on gnirehtet anymore (I still continue scrcpy, and both are on my free time only).

**Comment by q6654282** [comment]
All that remains is to wait! ha-ha

**Comment by manhtam922004** [comment]
Please fix it! Thank you so much!

**Comment by NA0341** [comment]
> All that remains is to wait! ha-ha


@q6654282 said:
> All that remains is to wait! ha-ha

I have no idea why you guys need to connect sooo many devices. And that to just one host device at that.

But: Why don't you just use multiple host devices?
I mean it can be a simple Single Board Computer (SBC) if you don't have other Computers around.
And for the network connection: use a simple, reliable switch. Done! 😉
